### ISSUE TITLE
rpc timeout during perform_tests call [blockdev_pmem->bdev_bounds]

### ISSUE THREAD

**Issue by mikeBashStuff** [spdk/spdk#1612]
```
SPDK: latest (v20.07-rc1-550-ge351dca37)
OS: Clearlinux (kernel: 5.4.56-56.lts2019, version: 33590)
```

Point of failure:
```
[2020-10-08 20:21:51.344894] Starting SPDK v20.10-pre git sha1 e351dca37 / DPDK 20.08.0 initialization...
[2020-10-08 20:21:51.345056] [ DPDK EAL parameters: [2020-10-08 20:21:51.345119] bdevio [2020-10-08 20:21:51.345157] --no-shconf [2020-10-08 20:21:51.345190] -c 0x7 [2020-10-08 20:21:51.345237] -m 0 [2020-10-08 20:21:51.345270] --log-level=lib.eal:6 [2020-10-08 20:21:51.345341] --log-level=lib.cryptodev:5 [2020-10-08 20:21:51.345376] --log-level=user1:6 [2020-10-08 20:21:51.345450] --iova-mode=pa [2020-10-08 20:21:51.345496] --base-virtaddr=0x200000000000 [2020-10-08 20:21:51.345531] --match-allocations [2020-10-08 20:21:51.345569] --file-prefix=spdk_pid62423 [2020-10-08 20:21:51.345625] ]
EAL: No available hugepages reported in hugepages-1048576kB
EAL: No legacy callbacks, legacy socket not created
[2020-10-08 20:21:51.476450] app.c: 667:spdk_app_start: *NOTICE*: Total cores available: 3
[2020-10-08 20:21:51.691346] reactor.c: 371:reactor_run: *NOTICE*: Reactor started on core 1
[2020-10-08 20:21:51.692953] reactor.c: 371:reactor_run: *NOTICE*: Reactor started on core 2
[2020-10-08 20:21:51.693957] reactor.c: 371:reactor_run: *NOTICE*: Reactor started on core 0
[2020-10-08 20:21:51.711291] accel_engine.c: 500:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
 20:21:52       # xtrace_enable
 20:21:52       # (( i == 0 ))
 20:21:52       # return 0
 20:21:52       # /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/tests.py perform_tests
I/O targets:
  Pmem0: 64700 blocks of 512 bytes (32 MiB)


     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/


Suite: bdevio tests on: Pmem0
  Test: blockdev write read 4k ...passed
  Test: blockdev write zeroes read 4k ...passed
  Test: blockdev write zeroes read 1m ...passed
  Test: blockdev write zeroes read 3m ...passed
  Test: blockdev write zeroes read 3.5m ...Timeout while waiting for response:


  20:22:52      # trap - ERR
  20:22:52      # print_backtrace
  20:22:52      # [[ ehxBET =~ e ]]
  20:22:52      # args=("${BASH_ARGV[@]}")
  20:22:52      # local args
  20:22:52      # xtrace_disable
  20:22:52      # '[' no '!=' yes ']'
  20:22:52      # PREV_BASH_OPTS=ehxBET
  20:22:52      # [[ ehxBET == *\x* ]]
  20:22:52      # XTRACE_DISABLED=yes
  20:22:52      # set +x
========== Backtrace start: ==========

in ./test/bdev/blockdev.sh:143 -> bdev_bounds([])
     ...
   138          $testdir/bdevio/bdevio -w -s $PRE_RESERVED_MEM --json "$conf_file" &
   139          bdevio_pid=$!
   140          trap 'killprocess $bdevio_pid; exit 1' SIGINT SIGTERM EXIT
   141          echo "Process bdevio pid: $bdevio_pid"
   142          waitforlisten $bdevio_pid
=> 143          $testdir/bdevio/tests.py perform_tests
   144          killprocess $bdevio_pid
   145          trap - SIGINT SIGTERM EXIT
   146  }
   147
   148  function nbd_function_test() {
     ...
in /home/vagrant/spdk_repo/spdk/test/common/autotest_common.sh:940 -> run_test(["bdev_bounds"],["bdev_bounds"])
     ...
   935          timing_enter $test_name
   936          echo "************************************"
   937          echo "START TEST $test_name"
   938          echo "************************************"
   939          xtrace_restore
=> 940          time "$@"
   941          xtrace_disable
   942          echo "************************************"
   943          echo "END TEST $test_name"
   944          echo "************************************"
   945          timing_exit $test_name
     ...
in ./test/bdev/blockdev.sh:366 -> main(["pmem"])
     ...
   361  killprocess "$spdk_tgt_pid"
   362  # End bdev configuration
   363  #-----------------------------------------------------
   364
   365  run_test "bdev_hello_world" $SPDK_EXAMPLE_DIR/hello_bdev --json "$conf_file" -b "$hello_world_bdev"
=> 366  run_test "bdev_bounds" bdev_bounds
   367  run_test "bdev_nbd" nbd_function_test $conf_file "$bdevs_name"
   368  if [[ $CONFIG_FIO_PLUGIN == y ]]; then
   369          if [ "$test_type" = "nvme" ] || [ "$test_type" = "gpt" ]; then
   370                  # TODO: once we get real multi-ns drives, re-enable this test for NVMe.
   371                  echo "skipping fio tests on NVMe due to multi-ns failures."
     ...

========== Backtrace end ==========
  20:22:52      # xtrace_enable
  20:22:52      # return 0

real    1m0.981s
user    0m0.348s
sys     0m0.078s
 20:22:52       # killprocess 62423
 20:22:52       # '[' -z 62423 ']'
 20:22:52       # kill -0 62423
  20:22:52      # uname
 20:22:52       # '[' Linux = Linux ']'
  20:22:52      # ps --no-headers -o comm= 62423
 20:22:52       # process_name=reactor_0
 20:22:52       # '[' reactor_0 = sudo ']'
 20:22:52       # echo 'killing process with pid 62423'
killing process with pid 62423
 20:22:52       # kill 62423
 20:22:52       # wait 62423
passed
  Test: blockdev reset ...passed
  Test: blockdev write read 512 bytes ...passed
  Test: blockdev write read size > 128k ...passed
  Test: blockdev write read invalid size ...passed
  Test: blockdev write read offset + nbytes == size of blockdev ...passed
  Test: blockdev write read offset + nbytes > size of blockdev ...passed
  Test: blockdev write read max offset ...passed
  Test: blockdev write read 8k on overlapped address offset ...passed
  Test: blockdev writev readv 4k ...passed
  Test: blockdev writev readv 30 x 4k ...passed
  Test: blockdev writev readv 512 bytes ...passed
  Test: blockdev writev readv size > 128k ...passed
  Test: blockdev writev readv size > 128k in two iovs ...passed
  Test: blockdev comparev and writev ...passed
  Test: blockdev nvme passthru rw ...passed
  Test: blockdev nvme passthru vendor specific ...passed
  Test: blockdev nvme admin passthru ...passed

Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites      1      1    n/a      0        0
               tests     22     22     22      0        0
             asserts     73     73     73      0      n/a

Elapsed time =   99.921 seconds
```

Then, ASAN trace followed for `bdevio`:
```
=================================================================
==62423==ERROR: AddressSanitizer: heap-use-after-free on address 0x7fcf90e05858 at pc 0x7fcfa2b92187 bp 0x7ffd2b5c5980 sp 0x7ffd2b5c5978
READ of size 8 at 0x7fcf90e05858 thread T0 (reactor_0)
    #0 0x7fcfa2b92186 in jsonrpc_server_send_response /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:344
    #1 0x7fcfa2b8ba59 in end_response /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:287
    #2 0x7fcfa2b8c1ab in spdk_jsonrpc_end_result /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:324
    #3 0x55a9f1c13f48 in rpc_perform_tests_cb /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1354
    #4 0x55a9f1c12d4e in __stop_init_thread /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1122
    #5 0x7fcfa2cdb498 in msg_queue_run_batch /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:505
    #6 0x7fcfa2cdcf3b in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:583
    #7 0x7fcfa2ce0896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #8 0x7fcfa2dda7e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #9 0x7fcfa2ddc07c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #10 0x7fcfa2ddd1a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #11 0x7fcfa2dd3d40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #12 0x55a9f1c1471d in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1430
    #13 0x7fcfa1901dbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)
    #14 0x55a9f1c0a86d in _start (/home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio+0x1086d)

0x7fcf90e05858 is located 32856 bytes inside of 2102832-byte region [0x7fcf90dfd800,0x7fcf90ffee30)
freed by thread T0 (reactor_0) here:
    #0 0x7fcfa45fee77 in free (/usr/lib64/libasan.so.6+0xcae77)
    #1 0x7fcfa2b8e4a0 in spdk_jsonrpc_server_shutdown /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:151
    #2 0x7fcfa2bd1fe5 in spdk_rpc_close /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:314
    #3 0x7fcfa2ddec54 in spdk_rpc_finish /home/vagrant/spdk_repo/spdk/lib/event/rpc.c:85
    #4 0x7fcfa2dd3eee in app_stop /home/vagrant/spdk_repo/spdk/lib/event/app.c:731
    #5 0x7fcfa2cdb498 in msg_queue_run_batch /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:505
    #6 0x7fcfa2cdcf3b in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:583
    #7 0x7fcfa2ce0896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #8 0x7fcfa2dda7e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #9 0x7fcfa2ddc07c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #10 0x7fcfa2ddd1a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #11 0x7fcfa2dd3d40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #12 0x55a9f1c1471d in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1430
    #13 0x7fcfa1901dbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

previously allocated by thread T0 (reactor_0) here:
    #0 0x7fcfa45ff337 in calloc (/usr/lib64/libasan.so.6+0xcb337)
    #1 0x7fcfa2b8c627 in spdk_jsonrpc_server_listen /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:46
    #2 0x7fcfa2bd12f7 in spdk_rpc_listen /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:198
    #3 0x7fcfa2ddebaa in spdk_rpc_initialize /home/vagrant/spdk_repo/spdk/lib/event/rpc.c:70
    #4 0x7fcfa2dcfee7 in app_start_rpc /home/vagrant/spdk_repo/spdk/lib/event/app.c:367
    #5 0x7fcfa2de2811 in app_json_config_load_done /home/vagrant/spdk_repo/spdk/lib/event/json_config.c:137
    #6 0x7fcfa2de5b7e in app_json_config_load_subsystem /home/vagrant/spdk_repo/spdk/lib/event/json_config.c:483
    #7 0x7fcfa2cdb498 in msg_queue_run_batch /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:505
    #8 0x7fcfa2cdcf3b in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:583
    #9 0x7fcfa2ce0896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #10 0x7fcfa2dda7e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #11 0x7fcfa2ddc07c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #12 0x7fcfa2ddd1a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #13 0x7fcfa2dd3d40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #14 0x55a9f1c1471d in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1430
    #15 0x7fcfa1901dbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

SUMMARY: AddressSanitizer: heap-use-after-free /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:344 in jsonrpc_server_send_response
Shadow bytes around the buggy address:
  0x0ffa721b8ab0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0ffa721b8ac0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0ffa721b8ad0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0ffa721b8ae0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0ffa721b8af0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0ffa721b8b00: fd fd fd fd fd fd fd fd fd fd fd[fd]fd fd fd fd
  0x0ffa721b8b10: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0ffa721b8b20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0ffa721b8b30: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0ffa721b8b40: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0ffa721b8b50: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==62423==ABORTING
  20:23:31      # trap - ERR
  20:23:31      # print_backtrace
  20:23:31      # [[ ehxBET =~ e ]]
  20:23:31      # args=("${BASH_ARGV[@]}")
  20:23:31      # local args
  20:23:31      # xtrace_disable
  20:23:31      # '[' no '!=' yes ']'
  20:23:31      # PREV_BASH_OPTS=ehxBET
  20:23:31      # [[ ehxBET == *\x* ]]
  20:23:31      # XTRACE_DISABLED=yes
  20:23:31      # set +x
========== Backtrace start: ==========

in /home/vagrant/spdk_repo/spdk/test/common/autotest_common.sh:940 -> run_test(["blockdev_pmem"],["./test/bdev/blockdev.sh"],["pmem"])
     ...
   935          timing_enter $test_name
   936          echo "************************************"
   937          echo "START TEST $test_name"
   938          echo "************************************"
   939          xtrace_restore
=> 940          time "$@"
   941          xtrace_disable
   942          echo "************************************"
   943          echo "END TEST $test_name"
   944          echo "************************************"
   945          timing_exit $test_name
     ...
in /home/vagrant/spdk_repo/spdk/autotest.sh:266 -> main(["/home/vagrant/autorun-spdk.conf"])
     ...
   261                  run_test "vhost_fuzz" ./test/vhost/fuzz/fuzz.sh
   262                  timing_exit vhost_initiator
   263          fi
   264
   265          if [ $SPDK_TEST_PMDK -eq 1 ]; then
=> 266                  run_test "blockdev_pmem" ./test/bdev/blockdev.sh "pmem"
   267                  run_test "pmem" ./test/pmem/pmem.sh -x
   268                  run_test "spdkcli_pmem" ./test/spdkcli/pmem.sh
   269          fi
   270
   271          if [ $SPDK_TEST_RBD -eq 1 ]; then
     ...

========== Backtrace end ==========
```
Spotted while running local tests via `autorun.sh` with the following config:

```
SPDK_RUN_VALGRIND=1
SPDK_RUN_FUNCTIONAL_TEST=1
SPDK_TEST_UNITTEST=1
SPDK_TEST_NVME=1
SPDK_TEST_BLOCKDEV=1
SPDK_TEST_NVMF=1
SPDK_TEST_PMDK=1
SPDK_TEST_JSON=1
SPDK_RUN_ASAN=1
SPDK_RUN_UBSAN=1
```

Can't figure out if `perform_tests()` is at fault here or `bdevio` so it's likely that the title of this issue should be changed. Any hints would be appreciated.

**Comment by yidong0635** [comment]
It seems there're two issues:
1. rpc can easily in a timeout state, there may be an issue in rpc running mechanism.
2. this time it triggers a heap-used-after-free issue.
     perform test `bdevio -w` :  set` g_wait_for_tests= true`. Due to the no response of rpc , connection was terminated. Timeout was not treated as a testing failure and then end response (`num_failures=0`) .

**Comment by Comphix** [comment]
Looks like when the bdev_bounds test completes, the killprocess send a signal to the App and trigger the default spdk_app_stop() operation in the default APP signal handling and the RPC server is freed. Then bdevio APP's own shutdown function called, where the g_wait_for_tests is set to TRUE and starts the RPC cleanup operation. At this moment, the RPC server structure is touched again which caused the heap-used-after-free issue and eventually caused the RPC timed out.

I'm also looking at the bdevperf implementation, where bdevperf's own shutdown function is defined and when the signal is sent, the bdevperf's shutdown function is called so as to properly cleanup.

In this case, for bdevio, the own shutdown function is also needed and at the same time, to have a g_shutdown flag to determine whether calling RPC function or app stop function with the same logic of bdevperf.

**Comment by Comphix** [comment]
Raised a patch here https://review.spdk.io/gerrit/c/spdk/spdk/+/4569 to check whether this issue is fixed or not.

**Comment by mikeBashStuff** [comment]
@Comphix Appreciate your help! :) I am afraid running tests with this patch introduced another issue - the rpc timeout is gone (sanitizer doesn't complain either), however, now `wait()` doesn't return while waiting for `bdevio` blocking the test for indefinite amount of time. This can be also seen on the CI side for jobs which run under this patch.

**Comment by tomzawadzki** [comment]
The patch resolving this issue, along with edit to pass CI [see previous comment] is now merged https://review.spdk.io/gerrit/c/spdk/spdk/+/4569.

**Comment by mikeBashStuff** [comment]
@tomzawadzki Issue is still seen on the latest master (including the mentioned fix). Failure occurs during the same test while executing `perform_tests()`. The sanitizer's trace differs though:
```

[2020-10-12 09:00:15.792244] Starting SPDK v20.10-pre git sha1 2881330db / DPDK 20.08.0 initialization...
[2020-10-12 09:00:15.792395] [ DPDK EAL parameters: [2020-10-12 09:00:15.792446] bdevio [2020-10-12 09:00:15.792492] --no-shconf [2020-10-12 09:00:15.792527] -c 0x7 [2020-10-12 09:00:15.792569] -m 0 [2020-10-12 09:00:15.792615] --log-level=lib.eal:6 [2020-10-12 09:00:15.792671] --log-level=lib.cryptodev:5 [2020-10-12 09:00:15.792720] --log-level=user1:6 [2020-10-12 09:00:15.792771] --iova-mode=pa [2020-10-12 09:00:15.792805] --base-virtaddr=0x200000000000 [2020-10-12 09:00:15.792839] --match-allocations [2020-10-12 09:00:15.792880] --file-prefix=spdk_pid1795317 [2020-10-12 09:00:15.793416] ]
EAL: No available hugepages reported in hugepages-1048576kB
EAL: No legacy callbacks, legacy socket not created
[2020-10-12 09:00:15.932983] app.c: 667:spdk_app_start: *NOTICE*: Total cores available: 3
[2020-10-12 09:00:16.154564] reactor.c: 371:reactor_run: *NOTICE*: Reactor started on core 1
[2020-10-12 09:00:16.156195] reactor.c: 371:reactor_run: *NOTICE*: Reactor started on core 2
[2020-10-12 09:00:16.157331] reactor.c: 371:reactor_run: *NOTICE*: Reactor started on core 0
[2020-10-12 09:00:16.174756] accel_engine.c: 500:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
 09:00:16       # xtrace_enable
 09:00:16       # (( i == 0 ))
 09:00:16       # return 0
 09:00:16       # /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/tests.py perform_tests
I/O targets:
  Pmem0: 64700 blocks of 512 bytes (32 MiB)


     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/


Suite: bdevio tests on: Pmem0
  Test: blockdev write read 4k ...passed
  Test: blockdev write zeroes read 4k ...passed
  Test: blockdev write zeroes read 1m ...passed
  Test: blockdev write zeroes read 3m ...passed
  Test: blockdev write zeroes read 3.5m ...Timeout while waiting for response:


  09:01:16      # trap - ERR
  09:01:16      # print_backtrace
  09:01:16      # [[ ehxBET =~ e ]]
  09:01:16      # args=("${BASH_ARGV[@]}")
  09:01:16      # local args
  09:01:16      # xtrace_disable
  09:01:16      # '[' no '!=' yes ']'
  09:01:16      # PREV_BASH_OPTS=ehxBET
  09:01:16      # [[ ehxBET == *\x* ]]
  09:01:16      # XTRACE_DISABLED=yes
  09:01:16      # set +x
========== Backtrace start: ==========

in ./test/bdev/blockdev.sh:143 -> bdev_bounds([])
     ...
   138          $testdir/bdevio/bdevio -w -s $PRE_RESERVED_MEM --json "$conf_file" &
   139          bdevio_pid=$!
   140          trap 'killprocess $bdevio_pid; exit 1' SIGINT SIGTERM EXIT
   141          echo "Process bdevio pid: $bdevio_pid"
   142          waitforlisten $bdevio_pid
=> 143          $testdir/bdevio/tests.py perform_tests
   144          killprocess $bdevio_pid
   145          trap - SIGINT SIGTERM EXIT
   146  }
   147
   148  function nbd_function_test() {
     ...
in /home/vagrant/spdk_repo/spdk/test/common/autotest_common.sh:940 -> run_test(["bdev_bounds"],["bdev_bounds"])
     ...
   935          timing_enter $test_name
   936          echo "************************************"
   937          echo "START TEST $test_name"
   938          echo "************************************"
   939          xtrace_restore
=> 940          time "$@"
   941          xtrace_disable
   942          echo "************************************"
   943          echo "END TEST $test_name"
   944          echo "************************************"
   945          timing_exit $test_name
     ...
in ./test/bdev/blockdev.sh:366 -> main(["pmem"])
     ...
   361  killprocess "$spdk_tgt_pid"
   362  # End bdev configuration
   363  #-----------------------------------------------------
   364
   365  run_test "bdev_hello_world" $SPDK_EXAMPLE_DIR/hello_bdev --json "$conf_file" -b "$hello_world_bdev"
=> 366  run_test "bdev_bounds" bdev_bounds
   367  run_test "bdev_nbd" nbd_function_test $conf_file "$bdevs_name"
   368  if [[ $CONFIG_FIO_PLUGIN == y ]]; then
   369          if [ "$test_type" = "nvme" ] || [ "$test_type" = "gpt" ]; then
   370                  # TODO: once we get real multi-ns drives, re-enable this test for NVMe.
   371                  echo "skipping fio tests on NVMe due to multi-ns failures."
     ...

========== Backtrace end ==========
  09:01:16      # xtrace_enable
  09:01:16      # return 0

real    1m1.011s
user    0m0.371s
sys     0m0.087s
 09:01:16       # killprocess 1795317
 09:01:16       # '[' -z 1795317 ']'
 09:01:16       # kill -0 1795317
  09:01:16      # uname
 09:01:16       # '[' Linux = Linux ']'
  09:01:16      # ps --no-headers -o comm= 1795317
 09:01:16       # process_name=reactor_0
 09:01:16       # '[' reactor_0 = sudo ']'
 09:01:16       # echo 'killing process with pid 1795317'
killing process with pid 1795317
 09:01:16       # kill 1795317
 09:01:16       # wait 1795317
=================================================================
==1795317==ERROR: AddressSanitizer: heap-use-after-free on address 0x6030000026c0 at pc 0x564415447320 bp 0x7f6d322f8530 sp 0x7f6d322f8528
READ of size 8 at 0x6030000026c0 thread T3 (reactor_2)
    #0 0x56441544731f in __blockdev_read /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:379
    #1 0x7f6d3858a498 in msg_queue_run_batch /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:505
    #2 0x7f6d3858bf3b in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:583
    #3 0x7f6d3858f896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #4 0x7f6d386897e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #5 0x7f6d3868b07c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #6 0x7f6d380b7f55 in eal_thread_loop ../lib/librte_eal/linux/eal_thread.c:127
    #7 0x7f6d3738a81e  (/usr/lib64/libpthread.so.0+0x981e)
    #8 0x7f6d372acfc2 in __clone (/usr/lib64/libc.so.6+0x123fc2)

0x6030000026c0 is located 16 bytes inside of 32-byte region [0x6030000026b0,0x6030000026d0)
freed by thread T0 (reactor_0) here:
    #0 0x7f6d39ec5e77 in free (/usr/lib64/libasan.so.6+0xcae77)
    #1 0x564415444289 in bdevio_cleanup_targets /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:190
    #2 0x56441544bcc4 in __stop_init_thread /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1120
    #3 0x7f6d3858a498 in msg_queue_run_batch /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:505
    #4 0x7f6d3858bf3b in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:583
    #5 0x7f6d3858f896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #6 0x7f6d386897e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #7 0x7f6d3868b07c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #8 0x7f6d3868c1a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #9 0x7f6d38682d40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #10 0x56441544d80f in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #11 0x7f6d371b0dbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

previously allocated by thread T0 (reactor_0) here:
    #0 0x7f6d39ec617f in __interceptor_malloc (/usr/lib64/libasan.so.6+0xcb17f)
    #1 0x564415443c67 in bdevio_construct_target /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:120
    #2 0x564415443f5c in bdevio_construct_targets /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:155
    #3 0x56441544d3b5 in rpc_perform_tests /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1395
    #4 0x7f6d3847ff5f in jsonrpc_handler /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:128
    #5 0x7f6d384400e9 in jsonrpc_server_handle_request /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:253
    #6 0x7f6d384381ee in parse_single_request /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:103
    #7 0x7f6d3843a093 in jsonrpc_parse_request /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:233
    #8 0x7f6d38440b07 in jsonrpc_server_conn_recv /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:315
    #9 0x7f6d38442692 in spdk_jsonrpc_server_poll /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:433
    #10 0x7f6d384803b8 in spdk_rpc_accept /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:217
    #11 0x7f6d3868db2e in rpc_subsystem_poll /home/vagrant/spdk_repo/spdk/lib/event/rpc.c:51
    #12 0x7f6d3858e9e5 in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:648
    #13 0x7f6d3858f896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #14 0x7f6d386897e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #15 0x7f6d3868b07c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #16 0x7f6d3868c1a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #17 0x7f6d38682d40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #18 0x56441544d80f in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #19 0x7f6d371b0dbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

Thread T3 (reactor_2) created by T0 (reactor_0) here:
    #0 0x7f6d39e3dc46 in __interceptor_pthread_create (/usr/lib64/libasan.so.6+0x42c46)
    #1 0x7f6d380aa4a8 in rte_eal_init ../lib/librte_eal/linux/eal.c:1244
    #2 0x7f6d382636d2 in spdk_env_init /home/vagrant/spdk_repo/spdk/lib/env_dpdk/init.c:568
    #3 0x7f6d38680f1b in app_setup_env /home/vagrant/spdk_repo/spdk/lib/event/app.c:530
    #4 0x7f6d386829bb in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:662
    #5 0x56441544d80f in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #6 0x7f6d371b0dbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

SUMMARY: AddressSanitizer: heap-use-after-free /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:379 in __blockdev_read
Shadow bytes around the buggy address:
  0x0c067fff8480: fa fa fd fd fd fa fa fa fd fd fd fd fa fa fd fd
  0x0c067fff8490: fd fa fa fa fd fd fd fa fa fa 00 00 04 fa fa fa
  0x0c067fff84a0: 00 00 00 fa fa fa fd fd fd fa fa fa fd fd fd fa
  0x0c067fff84b0: fa fa fd fd fd fd fa fa fd fd fd fa fa fa fd fd
  0x0c067fff84c0: fd fd fa fa fd fd fd fd fa fa fd fd fd fa fa fa
=>0x0c067fff84d0: fd fd fd fd fa fa fd fd[fd]fd fa fa fa fa fa fa
  0x0c067fff84e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff84f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8500: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8510: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8520: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==1795317==ABORTING

```

**Comment by Comphix** [comment]
Looks like below bdevio_cleanup_targets() has caused the issue.

Let me have a change. 

0x6030000026c0 is located 16 bytes inside of 32-byte region [0x6030000026b0,0x6030000026d0)
freed by thread T0 (reactor_0) here:
    #0 0x7f6d39ec5e77 in free (/usr/lib64/libasan.so.6+0xcae77)
    #1 0x564415444289 in bdevio_cleanup_targets /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:190
    #2 0x56441544bcc4 in __stop_init_thread /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1120
    #3 0x7f6d3858a498 in msg_queue_run_batch /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:505
    #4 0x7f6d3858bf3b in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:583
    #5 0x7f6d3858f896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #6 0x7f6d386897e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #7 0x7f6d3868b07c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #8 0x7f6d3868c1a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #9 0x7f6d38682d40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #10 0x56441544d80f in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #11 0x7f6d371b0dbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

**Comment by Comphix** [comment]
Raised a new patch https://review.spdk.io/gerrit/c/spdk/spdk/+/4604 and will check the CI result first. Locally I've run the related bdev_bounds test several rounds which is ok.

**Comment by mikeBashStuff** [comment]
Run the test again with https://review.spdk.io/gerrit/c/spdk/spdk/+/4604 on top and it still fails during `perform_tests`, however, sanitizer now detects mem leaks:
```
==2538875==ERROR: LeakSanitizer: detected memory leaks

Indirect leak of 32768 byte(s) in 1 object(s) allocated from:
    #0 0x7fedf823417f in __interceptor_malloc (/usr/lib64/libasan.so.6+0xcb17f)
    #1 0x7fedf67a6735 in jsonrpc_parse_request /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:196
    #2 0x7fedf67adb07 in jsonrpc_server_conn_recv /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:315
    #3 0x7fedf67af692 in spdk_jsonrpc_server_poll /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:433
    #4 0x7fedf67ed3b8 in spdk_rpc_accept /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:217
    #5 0x7fedf69fab2e in rpc_subsystem_poll /home/vagrant/spdk_repo/spdk/lib/event/rpc.c:51
    #6 0x7fedf68fb9e5 in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:648
    #7 0x7fedf68fc896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #8 0x7fedf69f67e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #9 0x7fedf69f807c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #10 0x7fedf69f91a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #11 0x7fedf69efd40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #12 0x55cef79b87fe in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #13 0x7fedf551ddbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

Indirect leak of 4136 byte(s) in 1 object(s) allocated from:
    #0 0x7fedf8234337 in calloc (/usr/lib64/libasan.so.6+0xcb337)
    #1 0x7fedf67517cf in spdk_json_write_begin /home/vagrant/spdk_repo/spdk/lib/json/json_write.c:79
    #2 0x7fedf67a694b in jsonrpc_parse_request /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:203
    #3 0x7fedf67adb07 in jsonrpc_server_conn_recv /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:315
    #4 0x7fedf67af692 in spdk_jsonrpc_server_poll /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:433
    #5 0x7fedf67ed3b8 in spdk_rpc_accept /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:217
    #6 0x7fedf69fab2e in rpc_subsystem_poll /home/vagrant/spdk_repo/spdk/lib/event/rpc.c:51
    #7 0x7fedf68fb9e5 in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:648
    #8 0x7fedf68fc896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #9 0x7fedf69f67e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #10 0x7fedf69f807c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #11 0x7fedf69f91a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #12 0x7fedf69efd40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #13 0x55cef79b87fe in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #14 0x7fedf551ddbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

Indirect leak of 128 byte(s) in 1 object(s) allocated from:
    #0 0x7fedf823417f in __interceptor_malloc (/usr/lib64/libasan.so.6+0xcb17f)
    #1 0x7fedf67a6303 in jsonrpc_parse_request /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:184
    #2 0x7fedf67adb07 in jsonrpc_server_conn_recv /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:315
    #3 0x7fedf67af692 in spdk_jsonrpc_server_poll /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:433
    #4 0x7fedf67ed3b8 in spdk_rpc_accept /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:217
    #5 0x7fedf69fab2e in rpc_subsystem_poll /home/vagrant/spdk_repo/spdk/lib/event/rpc.c:51
    #6 0x7fedf68fb9e5 in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:648
    #7 0x7fedf68fc896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #8 0x7fedf69f67e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #9 0x7fedf69f807c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #10 0x7fedf69f91a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #11 0x7fedf69efd40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #12 0x55cef79b87fe in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #13 0x7fedf551ddbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

Indirect leak of 88 byte(s) in 1 object(s) allocated from:
    #0 0x7fedf8234337 in calloc (/usr/lib64/libasan.so.6+0xcb337)
    #1 0x7fedf67a5b62 in jsonrpc_parse_request /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:161
    #2 0x7fedf67adb07 in jsonrpc_server_conn_recv /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:315
    #3 0x7fedf67af692 in spdk_jsonrpc_server_poll /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:433
    #4 0x7fedf67ed3b8 in spdk_rpc_accept /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:217
    #5 0x7fedf69fab2e in rpc_subsystem_poll /home/vagrant/spdk_repo/spdk/lib/event/rpc.c:51
    #6 0x7fedf68fb9e5 in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:648
    #7 0x7fedf68fc896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #8 0x7fedf69f67e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #9 0x7fedf69f807c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #10 0x7fedf69f91a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #11 0x7fedf69efd40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #12 0x55cef79b87fe in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #13 0x7fedf551ddbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

Indirect leak of 63 byte(s) in 1 object(s) allocated from:
    #0 0x7fedf823417f in __interceptor_malloc (/usr/lib64/libasan.so.6+0xcb17f)
    #1 0x7fedf67a5e3b in jsonrpc_parse_request /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server.c:172
    #2 0x7fedf67adb07 in jsonrpc_server_conn_recv /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:315
    #3 0x7fedf67af692 in spdk_jsonrpc_server_poll /home/vagrant/spdk_repo/spdk/lib/jsonrpc/jsonrpc_server_tcp.c:433
    #4 0x7fedf67ed3b8 in spdk_rpc_accept /home/vagrant/spdk_repo/spdk/lib/rpc/rpc.c:217
    #5 0x7fedf69fab2e in rpc_subsystem_poll /home/vagrant/spdk_repo/spdk/lib/event/rpc.c:51
    #6 0x7fedf68fb9e5 in thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:648
    #7 0x7fedf68fc896 in spdk_thread_poll /home/vagrant/spdk_repo/spdk/lib/thread/thread.c:691
    #8 0x7fedf69f67e1 in _reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:326
    #9 0x7fedf69f807c in reactor_run /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:382
    #10 0x7fedf69f91a4 in spdk_reactors_start /home/vagrant/spdk_repo/spdk/lib/event/reactor.c:477
    #11 0x7fedf69efd40 in spdk_app_start /home/vagrant/spdk_repo/spdk/lib/event/app.c:711
    #12 0x55cef79b87fe in main /home/vagrant/spdk_repo/spdk/test/bdev/bdevio/bdevio.c:1439
    #13 0x7fedf551ddbc in __libc_start_main (/usr/lib64/libc.so.6+0x27dbc)

SUMMARY: AddressSanitizer: 37183 byte(s) leaked in 5 allocation(s).
```
Not sure if this still relates directly to `bdevio` though.

**Comment by mikeBashStuff** [comment]
@Comphix I hit this particular trace on clearlinux VM (`VERSION_ID=33590`)  by running `autorun.sh` with this config (the same way I triggered two previous instances of this issue):
```
SPDK_RUN_VALGRIND=1
SPDK_RUN_FUNCTIONAL_TEST=1
SPDK_TEST_UNITTEST=1
SPDK_TEST_NVME=1
SPDK_TEST_BLOCKDEV=1
SPDK_TEST_NVMF=1
SPDK_TEST_PMDK=1
SPDK_TEST_JSON=1
SPDK_RUN_ASAN=1
SPDK_RUN_UBSAN=1
```
I also cherry-picked your commit on top of `v20.07-rc1-566-gb866d5192` and run the test with it included.

**Comment by yidong0635** [comment]
since our test environment was damaged, I will install a new one to check and verify Gang's idea.

**Comment by Comphix** [comment]
@mikeBashStuff , could you help to get the latest change from below patch and have another try as @yidong0635 is helping to set the same clearlinux OS now? Many thanks for your time and help.

https://review.spdk.io/gerrit/c/spdk/spdk/+/4604

**Comment by mikeBashStuff** [comment]
@Comphix Apologies for the late reply. With the 4th patchset from https://review.spdk.io/gerrit/c/spdk/spdk/+/4604 I am afraind I am getting the same results as in https://github.com/spdk/spdk/issues/1612#issuecomment-707650993.

**Comment by yidong0635** [comment]
I can't reproduce such error until now. 
using commit:`779a6bdf82e3746036503c715770799b038ebce7` , 

and it blocked by #1624 
```
NAME="Clear Linux OS"
VERSION=1
ID=clear-linux-os
ID_LIKE=clear-linux-os
VERSION_ID=33420
gcc version 10.1.1 20200619 releases/gcc-10.1.0-223-gf8e16e23d0 (Clear Linux OS for Intel Architecture)

```

**Comment by mikeBashStuff** [comment]
Just to note, I've tested it on `33420` release, and the issue is present there as well I am afraid.

**Comment by Comphix** [comment]
@mikeBashStuff , thanks. So what's the Version ID of the clearlinux that you are using to hit the issue described here?

**Comment by yidong0635** [comment]
I can reproduce a similar issue sometimes. (memory leaks)
at `run_test "json_config" ./test/json_config/json_config.sh`
[issue.txt](https://github.com/spdk/spdk/files/5389134/issue.txt)

**Comment by yidong0635** [comment]
my issue is caused by:
```
Waiting for process to start up and listen on UNIX domain socket /var/tmp/spdk_tgt.sock...
 06:20:09       # xtrace_disable
 06:20:09       # '[' no '!=' yes ']'
/home/clear/spdk/build/bin/spdk_tgt: error while loading shared libraries: libspdk_bdev_malloc.so.2.0: cannot open shared object file: No such file or directory
 06:20:09       # PREV_BASH_OPTS=ehxBET
 06:20:09       # [[ ehxBET == *\x* ]]
 06:20:09       # XTRACE_DISABLED=yes
 06:20:09       # set +x
/home/clear/spdk/test/common/autotest_common.sh: line 660: kill: (82850) - No such process
ERROR: process (pid: 82850) is no longer running
 06:20:09       # xtrace_enable

```

**Comment by yidong0635** [comment]
`waitforlisten`  cannot found sock, why don't exit and continue to run test? this due to some errors. @mikeBashStuff

**Comment by mikeBashStuff** [comment]
@Comphix I've managed to reproduce it on `33420`, `33590` and `33830` so far. Though, considering how Clearlinux's build-cadence works, these versions shouldn't differ that much really (e.g., the most notable difference I noticed between `33830` and the rest is that `gcc` was upgraded to `10.2.1`). If it helps, I can upload somewhere a vagrant box I build for `clearlinux` (or raw disk image as an alternative) to help you with debugging (that way we can be sure we are testing it on exactly the same snapshot of the distro). Though it's quite huge, so the question is if I could put it on some system for you to pick up. Alternatively, I can just send it via same filesharing platform if that's ok. 

@yidong0635 `waitfolisten()` is meant to determine if app that was just executed is up and running. If it isn't, we should actually fail since it means we hit some abnormal scenario. In your example, it's actually what I would expect, considering this:
`/home/clear/spdk/build/bin/spdk_tgt: error while loading shared libraries: libspdk_bdev_malloc.so.2.0: cannot open shared object file: No such file or directory`.

**Comment by yidong0635** [comment]
Hi all,
My configure:
```
PDK_RUN_VALGRIND=1
SPDK_RUN_FUNCTIONAL_TEST=1
SPDK_TEST_UNITTEST=0
SPDK_TEST_NVME=1
SPDK_TEST_BLOCKDEV=1
SPDK_TEST_PMDK=1
SPDK_TEST_NVMF=1
SPDK_TEST_JSON=1
SPDK_RUN_ASAN=1
SPDK_RUN_UBSAN=1
```

```

************************************
************************************
START TEST bdev_bounds
************************************
Process bdevio pid: 380637
Waiting for process to start up and listen on UNIX domain socket /var/tmp/spdk.sock...
EAL: No legacy callbacks, legacy socket not created
I/O targets:
  Pmem0: 64700 blocks of 512 bytes (32 MiB)


     CUnit - A unit testing framework for C - Version 2.1-3
     http://cunit.sourceforge.net/


Suite: bdevio tests on: Pmem0
  Test: blockdev write read 4k ...passed
  Test: blockdev write zeroes read 4k ...passed
  Test: blockdev write zeroes read 1m ...passed
  Test: blockdev write zeroes read 3m ...passed
  Test: blockdev write zeroes read 3.5m ...passed
  Test: blockdev reset ...passed
  Test: blockdev write read 512 bytes ...passed
  Test: blockdev write read size > 128k ...passed
  Test: blockdev write read invalid size ...passed
  Test: blockdev write read offset + nbytes == size of blockdev ...passed
  Test: blockdev write read offset + nbytes > size of blockdev ...passed
  Test: blockdev write read max offset ...passed
  Test: blockdev write read 8k on overlapped address offset ...passed
  Test: blockdev writev readv 4k ...passed
  Test: blockdev writev readv 30 x 4k ...passed
  Test: blockdev writev readv 512 bytes ...passed
  Test: blockdev writev readv size > 128k ...passed
  Test: blockdev writev readv size > 128k in two iovs ...passed
  Test: blockdev comparev and writev ...passed
  Test: blockdev nvme passthru rw ...passed
  Test: blockdev nvme passthru vendor specific ...passed
  Test: blockdev nvme admin passthru ...passed

Run Summary:    Type  Total    Ran Passed Failed Inactive
              suites      1      1    n/a      0        0
               tests     22     22     22      0        0
             asserts     73     73     73      0      n/a

Elapsed time =   21.323 seconds
0
killing process with pid 380637
************************************
END TEST bdev_bounds
************************************
************************************
START TEST bdev_nbd
************************************
Process nbd pid: 380683
Waiting for process to start up and listen on UNIX domain socket /var/tmp/spdk-nbd.sock...
EAL: No legacy callbacks, legacy socket not created
killing process with pid 380683
************************************
END TEST bdev_nbd
************************************
************************************
START TEST bdev_fio
************************************
************************************
START TEST bdev_fio_rw_verify
************************************
EAL: No legacy callbacks, legacy socket not created
************************************
END TEST bdev_fio_rw_verify
************************************
************************************
START TEST bdev_fio_trim
************************************
EAL: No legacy callbacks, legacy socket not created
************************************
END TEST bdev_fio_trim
************************************
************************************
END TEST bdev_fio
************************************
************************************
START TEST bdev_verify
************************************
EAL: No legacy callbacks, legacy socket not created
Running I/O for 5 seconds...
^M Thread name: Pmem0_1
^M Core Mask: 0x1
^M Pmem0               :      25.60 IOPS       0.10 MiB/s
^M Thread name: Pmem0_2
^M Core Mask: 0x2
^M Pmem0               :      25.60 IOPS       0.10 MiB/s
^M =====================================================
^M Total               :      51.20 IOPS       0.20 MiB/s
************************************
END TEST bdev_verify
************************************
************************************
START TEST bdev_write_zeroes
************************************
EAL: No legacy callbacks, legacy socket not created
Running I/O for 1 seconds...
^M Thread name: Pmem0_1
^M Core Mask: 0x1
^M Pmem0               :     136.00 IOPS       0.53 MiB/s
^M =====================================================
^M Total               :     136.00 IOPS       0.53 MiB/s
************************************
END TEST bdev_write_zeroes
************************************
************************************
END TEST blockdev_pmem
************************************

```
```
root@clearlinux/home/clear/spdk # cat /etc/os-release
NAME="Clear Linux OS"
VERSION=1
ID=clear-linux-os
ID_LIKE=clear-linux-os
VERSION_ID=33840
PRETTY_NAME="Clear Linux OS"
ANSI_COLOR="1;35"
HOME_URL="https://clearlinux.org"
SUPPORT_URL="https://clearlinux.org"
BUG_REPORT_URL="mailto:dev@lists.clearlinux.org"
PRIVACY_POLICY_URL="http://www.intel.com/privacy"
BUILD_ID=33840
```
```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib64/gcc/x86_64-generic-linux/10/lto-wrapper
Target: x86_64-generic-linux
Configured with: ../gcc-10.2.0/configure --prefix=/usr --with-pkgversion='Clear Linux OS for Intel Architecture' --libdir=/usr/lib64 --enable-libstdcxx-pch --libexecdir=/usr/lib64 --with-system-zlib --enable-shared --enable-gnu-indirect-function --disable-vtable-verify --enable-threads=posix --enable-__cxa_atexit --enable-plugin --enable-ld=default --enable-clocale=gnu --disable-multiarch --enable-multilib --enable-lto --disable-werror --enable-linker-build-id --build=x86_64-generic-linux --target=x86_64-generic-linux --enable-languages=c,c++,fortran,go --enable-bootstrap --with-ppl=yes --with-isl --includedir=/usr/include --exec-prefix=/usr --with-glibc-version=2.19 --disable-libunwind-exceptions --with-gnu-ld --with-tune=haswell --with-arch=westmere --enable-cet --disable-libmpx --with-gcc-major-version-only --enable-default-pie
Thread model: posix
Supported LTO compression algorithms: zlib
gcc version 10.2.1 20201014 releases/gcc-10.2.0-375-gb6bf8afea2 (Clear Linux OS for Intel Architecture)

commit 10f712bc99e472657555ab7e9d9f0565db528ae8
```

And 
```
root@clearlinux/home/clear/spdk # free -h
              total        used        free      shared  buff/cache   available
Mem:          9.8Gi       8.2Gi       1.3Gi       0.0Ki       368Mi       1.4Gi
Swap:         254Mi        16Mi       238Mi

root@clearlinux/home/clear/spdk # echo $(nproc)
16
```

**Comment by Comphix** [comment]
Thanks @yidong0635 , just confirm that in your environment and with my patch, this issue is not reproduced? What's the different between your environment and Michal's? Thanks, Gang.

**Comment by yidong0635** [comment]
I just  used the commit : 10f712bc99e472657555ab7e9d9f0565db528ae8
without applying patch to test autorun.sh.  I followed the Michal's description to update OS and GCC.
And I test several times, It's nothing reported related to the title .

**Comment by tomzawadzki** [comment]
[Bug scrub] @mikeBashStuff Just to clear up, does the failure occur with 100% reproduction ?

**Comment by mikeBashStuff** [comment]
@tomzawadzki Yes, at least on the clearlinux builds I am working on (i.e. the ones built under our packer work).

**Comment by mikeBashStuff** [comment]
Ok, I run some additional tests and here are my findings:

I deployed the very same Clearlinux VM on a bit different system (hardware-wise, and distro-wise) and, to my surprise, issue was gone - I run the VM with the qemu shipped with the new distro (3.0.1, Fedora29), issue didn't appear. Then I compiled 3.1.0 qemu with series of debian patches (51) to align it with the one from original system I was using. Again, couldn't reproduce it. I adjusted cpu topology, including the model (number of supported instructions was still a bit different, but overall, almost the same), to match the VM setup from the original system. Couldn't reproduce.  Compiled and tested against vanilla 4.2.1 qemu on both systems - issue still present on the original systems, couldn't reproduce on the new one.

At the moment, the only difference is that on the original, affected system, the image of the nvme drive which is attached to the VM is stored on a plain rotational hdd. The performance of the VM is visibly lower there (tests take more time to finish, VM boots slower, IO in general feels clunky etc.) - on the second system ssd is used for all the images. So, considering the nature of the issue which seems to be affecting `perform_tests()` (sudden timeout), maybe it's actually corelated with poor storage the VM is run on top of.

Long story short, since this cannot be easily reproduced, let's close this issue. I can share the core dump from the `bdevio` and ASAN traces if anyone would be interested in troubleshooting this further, but as of now, I think this one can be closed.
